R/cube-transforms.R

```diff
@@ -86,37 +118,187 @@ setMethod("subtotalArray", "CrunchCube", function(x, headings = FALSE) {
 #'
 #' @aliases subtotalArray
 #' @export
-applyTransforms <- function (x, array = cubeToArray(x), ...) {
-    # if any of the dimensions are subvariables, don't even attempt to calculate
-    # transforms or insertions.
-    if (any(getDimTypes(x) == "ca_items")) {
-        return(array)
-    }
+applyTransforms <- function (x,
+                             array = cubeToArray(showMissing(x)),
+                             transforms_list = transforms(x),
+                             dims_list = dimensions(x),
+                             useNA = x@useNA,
+                             ...) {
+    trans_indices <- which(vapply(transforms_list, Negate(is.null), logical(1)))
+    
+    # Try to calculate the transforms for any dimensiton that have them, but
+    # fail silently if they aren't calcuable. We use a for loop so that failing
+    # one dimension doesn't break others. We could possibly use something like
+    # abind::abind here and bind together the insertion vectors in array form,
+    # but that would add a dependency
+    for (d in trans_indices) {
+        try({
```

Comment:
Why so much `try()` in here?

Reply:
I added the `try()`s here to prevent a malformed (or unanticipated) transform from preventing the cube from being displayed or being accessible.

Reply:
I'm curious how the `try()`s might interact. You have a for loop and a try within each iteration, then a try around something else at the bottom. If one of my loop iterations errors, do they all error? Do they leave `array` in a bad state? Will the next `try` necessarily also catch an error? 

Are there real cases where you can trigger these errors that you're catching? Are they tested?

I'm wondering whether it might better to have one big tryCatch around everything that on error just returns `array` unmodified.

Reply:
The `try()` inside the loop is an attempt to provide transforms for dims that have valid transforms. I've added a few tests for that condition now, and it does work as expected (and in the process caught a problem with the `prettyPrint` code that was preventing them from being formatted). 

I've removed the second try because I don't think it's necessary now. (It wasn't being triggered by failures above, at all.)

That being said, this is very much an edge case that I'm not sure anyone would hit, so maybe we do just want to wrap the whole thing in `tryCatch()` and call it a day.